pkg/kubelet/eviction/memory_threshold_notifier.go

```diff
@@ -0,0 +1,99 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eviction
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	statsapi "k8s.io/kubernetes/pkg/kubelet/apis/stats/v1alpha1"
+	"k8s.io/kubernetes/pkg/kubelet/cm"
+	evictionapi "k8s.io/kubernetes/pkg/kubelet/eviction/api"
+)
+
+const memoryUsageAttribute = "memory.usage_in_bytes"
+
+type memoryThresholdNotifier struct {
+	notifier   ThresholdNotifier
+	threshold  evictionapi.Threshold
+	cgroupPath string
+}
+
+// NewMemoryThresholdNotifier creates a MemoryThresholdNotifier which is designed to respond to the given threshold.
+// UpdateThreshold must be called once before the threshold will be active.
+func NewMemoryThresholdNotifier(threshold evictionapi.Threshold, cgroupRoot string, handler func(string)) (MemoryThresholdNotifier, error) {
+	cgroups, err := cm.GetCgroupSubsystems()
+	if err != nil {
+		return nil, err
+	}
+	cgpath, found := cgroups.MountPoints["memory"]
+	if !found || len(cgpath) == 0 {
+		return nil, fmt.Errorf("memory cgroup mount point not found")
+	}
+	if isAllocatableEvictionThreshold(threshold) {
+		cgpath += cgroupRoot
+	}
+	memoryNotifier := &memoryThresholdNotifier{
+		threshold:  threshold,
+		cgroupPath: cgpath,
+	}
+	message := fmt.Sprintf("eviction manager: %s crossed", memoryNotifier.Description())
+	memoryNotifier.notifier = NewThresholdNotifier(message, handler)
+	glog.Infof("eviction manager: created %s", memoryNotifier.Description())
+	return memoryNotifier, nil
+}
+
+func (m *memoryThresholdNotifier) UpdateThreshold(summary *statsapi.Summary) error {
+	memoryStats := summary.Node.Memory
+	if isAllocatableEvictionThreshold(m.threshold) {
+		allocatableContainer, err := getSysContainer(summary.Node.SystemContainers, statsapi.SystemContainerPods)
+		if err != nil {
+			return err
+		}
+		memoryStats = allocatableContainer.Memory
+	}
+	if memoryStats == nil || memoryStats.UsageBytes == nil || memoryStats.WorkingSetBytes == nil || memoryStats.AvailableBytes == nil {
+		return fmt.Errorf("summary was incomplete")
+	}
+	// Set threshold on usage to capacity - eviction_hard + inactive_file,
+	// since we want to be notified when working_set = capacity - eviction_hard
+	inactiveFile := resource.NewQuantity(int64(*memoryStats.UsageBytes-*memoryStats.WorkingSetBytes), resource.BinarySI)
+	capacity := resource.NewQuantity(int64(*memoryStats.AvailableBytes+*memoryStats.WorkingSetBytes), resource.BinarySI)
+	evictionThresholdQuantity := evictionapi.GetThresholdQuantity(m.threshold.Value, capacity)
+	memcgThreshold := capacity.DeepCopy()
+	memcgThreshold.Sub(*evictionThresholdQuantity)
+	memcgThreshold.Add(*inactiveFile)
+	threshold := strconv.FormatInt(memcgThreshold.Value(), 10)
+	glog.V(3).Infof("eviction manager: setting %s to %s\n", m.Description(), threshold)
+	return m.notifier.UpdateThreshold(m.cgroupPath, memoryUsageAttribute, threshold)
```

Comment:
Would it be better for UpdateThreshold to take a numerical argument for the threshold, instead of a string (potentially more type-safe)?

Reply:
done